### ISSUE TITLE
Interpolation of data is wrong when asked for a time before the first provided time

### ISSUE THREAD

**Issue by spco** [AtChem/AtChem2#294]
The current behaviour in `getConstrainedQuantAtT()` is this: if the required time is within the time covered by the constraint data, then it interpolates. Otherwise, it returns the value of the _last_ constraint data.

This certainly needs fixing in at least one case: if the required time is _before_ the first constraint data time, then the _last_ constraint data value is still returned. This is obviously not what is wanted.

As mentioned in 1. in #270, we may want to ensure all times are covered by constraint data. Or alternatively we might want to take the approach that all values before the first constraint data use that first value, and all after the last constraint data use the last value.

Which of these should we do? The latter is only a couple of lines' change from where we are.

The former would require us to check each of the constraints to make sure they cover all the times that the model covers. One caveat in this case: CVODE doesn't guarantee to only compute up to the final time we ask for. Instead, it steps until it reaches at least that time, then interpolates back if necessary. 

i.e. If we want the model to go from `t=0 to 1`, CVODE may calculate the values at `t=0.6, 1.2`, and then interpolate between those two to read out the value at `t=1`. While that is an extreme example (it's normally a lot smaller overshoot than that), that's what happens. CVODE then needs to be able to evaluate the constraint data at `t=1.2`. Thus we need to ensure the constraint data always covers that extra time for overshoot (which I don't think we can know a priori). We can either do that by demanding that data is given for that extra overshoot time, or we do it by demanding data up to `t=1`, and then extrapolate the data as necessary for the overshoot beyond that (either use the final data value, or extrapolate in a more clever way with linear/quad interpolation or splines etc).

Either are do-able - which is preferable?

**Comment by rs028** [comment]
Need to think about this one, it looks like it has the potential of creating numerical artifacts, if not errors. At some point all the interpolation-related stuff will need a close look. Either we do this now (and maybe postpone the function parser implementation) or we keep it for later.
Could it also have impact on the numerical issues we were talking about? That may help decide what to prioritize.

**Comment by spco** [comment]
I don't think this should have any particular effects on the numerical issues, as the interpolation is largely sound (though the use of left-handed constant interpolation will generally not do numerical stability many favours), as is the behaviour of CVODE. The incorrect use of the last value if the requested time is earlier than the supplied interval is definitely a bug, but the rest is more a case of "how do we want this to work?".

The automatic checking that all constraints supply enough data is almost a separate question, but is related.

**Comment by rs028** [comment]
Okay. I am labeling this as a bug and we can address it later. There are a number of things I think need to be done with the interpolation, and I suspect they are tied to the speed of the model integration and also its accuracy. So this could become big.

For the moment it remains responsibility of the user to make sure the constraint files have enough data.

**Comment by spco** [comment]
To summarise our previous discussions on this point:
(1) this isn't a 'numerical issue' as such, but is a bug
(2) as long as the user supplies sufficient data, then this will be absolutely fine. It only triggers in the case where the user asks to constrain a given species but doesn't supply the necessary data at t=0.

This can be fixed by either 
(a) using the first-supplied value for the species for any times before the first timestep for which it _is_ supplied; or
(b) using the initial value of the species up until the constraint data is supplied; or
(c) interpolating between the initial value and the first constrained data supplied; or
(d) check at startup whether all the necessary data has been supplied, and complain if it hasn't been.
